src/main/java/ch/njol/skript/expressions/ExprBarterOutput.java

```diff
@@ -0,0 +1,167 @@
+ *  along with Skript.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Copyright Peter Güttinger, SkriptLang team and contributors
+ */
+package ch.njol.skript.expressions;
+
+import ch.njol.skript.Skript;
+import ch.njol.skript.aliases.ItemType;
+import ch.njol.skript.classes.Changer;
+import ch.njol.skript.doc.Description;
+import ch.njol.skript.doc.Examples;
+import ch.njol.skript.doc.Name;
+import ch.njol.skript.doc.Since;
+import ch.njol.skript.lang.Expression;
+import ch.njol.skript.lang.ExpressionType;
+import ch.njol.skript.lang.SkriptParser.ParseResult;
+import ch.njol.skript.lang.util.SimpleExpression;
+import ch.njol.skript.log.ErrorQuality;
+import ch.njol.util.Kleenean;
+import ch.njol.util.coll.CollectionUtils;
+import org.bukkit.event.Event;
+import org.bukkit.event.entity.PiglinBarterEvent;
+import org.bukkit.inventory.ItemStack;
+import org.eclipse.jdt.annotation.Nullable;
+
+import java.util.ArrayList;
+import java.util.List;
+
+@Name("Barter Output")
+@Description("The items dropped by the piglin in a piglin bartering event.")
+@Examples({"on piglin barter:",
+			"\tif the bartering output contains a jack-o-lantern:",
+			"\t\tremove jack-o-lantern from bartering output",
+			"\t\tbroadcast \"Get out your amulets!\""})
+@Since("INSERT VERSION")
+public class ExprBarterOutput extends SimpleExpression<ItemType> {
+	
+	static {
+		Skript.registerExpression(ExprBarterOutput.class, ItemType.class,
+			ExpressionType.SIMPLE, "[the] [piglin] barter[ing] out(come|put)");
+	}
+
+	private Kleenean delay;
+
+	@Override
+	public boolean init(Expression<?>[] exprs, int matchedPattern,
+						Kleenean isDelayed, ParseResult parser) {
+		if (!getParser().isCurrentEvent(PiglinBarterEvent.class)) {
+			Skript.error("The expression 'barter input' can only be used in the piglin bartering event", ErrorQuality.SEMANTIC_ERROR);
+			return false;
+		}
+
+		delay = isDelayed;
+
+		return true;
+	}
+	
+	@Override
+	@Nullable
+	protected ItemType[] get(Event event) {
+		if (!(event instanceof PiglinBarterEvent)) {
+			return null;
+		}
+
+		return ((PiglinBarterEvent) event).getOutcome()
+			.stream()
+			.map(ItemType::new)
+			.toArray(ItemType[]::new);
+	}
+
+	@Override
+	@Nullable
+	public Class<?>[] acceptChange(Changer.ChangeMode mode) {
+		if (delay != Kleenean.FALSE) {
+			Skript.error("Can't change the piglin bartering outcome after the event has already passed", ErrorQuality.SEMANTIC_ERROR);
+			return null;
+		}
+
+		switch (mode) {
+			case SET:
+			case ADD:
+			case REMOVE:
+			case REMOVE_ALL:
+			case DELETE:
+				return CollectionUtils.array(ItemType[].class);
+			default:
+				return null;
+		}
+	}
+
+	@Override
+	public void change(Event event, @Nullable Object[] delta, Changer.ChangeMode mode) {
+		if (!(event instanceof PiglinBarterEvent)) {
+			return;
+		}
+
+		List<ItemType> deltaOutput = new ArrayList<>();
+
+		if (delta != null) {
+			for (Object deltaDrop : delta) {
+				if (deltaDrop instanceof ItemType) {
```

Comment:
this is technically an unnecessary check, proper api use would ensure that delta is ItemType because that's the only type you returned in acceptChange

Reply:
i feel like this section here would be better as a stream

Reply:
It would not but that's okay

Reply:
tbh it doesn't really need to exist at all

Reply:
ah, thanks for the clarification